File: sdk/typescript/src/client/databases.ts
import { listQuery, type HttpTransport } from "../http.js";
import type {
  Change,
  Document,
  DocumentListParams,
  ListChangesResponse,
  ListMeta,
  UpdateDocumentInput,
  UpsertDocumentInput,
} from "../types.js";
import type { QueryAst } from "../query.js";

export class ClientDatabasesService {
  constructor(private readonly http: HttpTransport) {}

  async createDocument(
    databaseId: string,
    collectionId: string,
    input: {
      document_id?: string;
      data: Record<string, unknown>;
      permissions?: string[];
    }
  ): Promise<Document> {
    return this.http.request<Document>(
      "POST",
      `/v1/databases/${databaseId}/collections/${collectionId}/documents`,
      {
        body: {
          database_id: databaseId,
          collection_id: collectionId,
          document_id: input.document_id ?? "",
          data: input.data,
          permissions: input.permissions,
        },
      }
    );
  }

  // C7 单 AST：带 query（过滤/排序/投影）时走 POST :list（body 即 Query
  // JSON，分页字段并入）；无 query 时保留 GET 简单分页（page_size/page_token）。
  async listDocuments(
    databaseId: string,
    collectionId: string,
    params?: DocumentListParams
  ): Promise<{ documents: Document[]; meta?: ListMeta }> {
    if (params?.query) {
      const body = documentsQueryBody(params.query, params.page_size, params.page_token);
      const res = await this.http.request<{ documents: Document[]; meta?: ListMeta }>(
        "POST",
        `/v1/databases/${databaseId}/collections/${collectionId}/documents:list`,
        { body }
      );
      return { documents: res.documents ?? [], meta: res.meta };
    }
    const res = await this.http.request<{
      documents: Document[];
      meta?: ListMeta;
    }>("GET", `/v1/databases/${databaseId}/collections/${collectionId}/documents`, {
      query: listQuery(params),
    });
    return { documents: res.documents ?? [], meta: res.meta };
  }

  async getDocument(
    databaseId: string,
    collectionId: string,
    documentId: string
  ): Promise<Document> {
    return this.http.request<Document>(
      "GET",
      `/v1/databases/${databaseId}/collections/${collectionId}/documents/${documentId}`
    );
  }

  async updateDocument(
    databaseId: string,
    collectionId: string,
    documentId: string,
    input: UpdateDocumentInput
  ): Promise<Document> {
    return this.http.request<Document>(
      "PATCH",
      `/v1/databases/${databaseId}/collections/${collectionId}/documents/${documentId}`,
      {
        body: {
          database_id: databaseId,
          collection_id: collectionId,
          document_id: documentId,
          ...input,
        },
      }
    );
  }

  async upsertDocument(
    databaseId: string,
    collectionId: string,
    documentId: string,
    input: UpsertDocumentInput
  ): Promise<Document> {
    return this.http.request<Document>(
      "PUT",
      `/v1/databases/${databaseId}/collections/${collectionId}/documents/${documentId}`,
      {
        body: {
          database_id: databaseId,
          collection_id: collectionId,
          document_id: documentId,
          ...input,
        },
      }
    );
  }

  async deleteDocument(
    databaseId: string,
    collectionId: string,
    documentId: string,
    // 用户集合 OCC 必填：来自上一次 GetDocument/List 的 version。
    version: number
  ): Promise<void> {
    await this.http.request<void>(
      "DELETE",
      `/v1/databases/${databaseId}/collections/${collectionId}/documents/${documentId}`,
      { query: { version } }
    );
  }

  // C7 单 AST：带 query 过滤时走 POST :count（GET 面仅支持无过滤计数）。
  async countDocuments(
    databaseId: string,
    collectionId: string,
    params?: Pick<DocumentListParams, "query">
  ): Promise<string | number> {
    if (params?.query) {
      const res = await this.http.request<{ count: string | number }>(
        "POST",
        `/v1/databases/${databaseId}/collections/${collectionId}/documents:count`,
        {
          body: {
            database_id: databaseId,
            collection_id: collectionId,
            query: params.query,
          },
        }
      );
      return res.count ?? 0;
    }
    // int64：网关序列化为字符串（如 "42"）；消费时建议 Number() 归一化。
    // GET 面只做无过滤计数；有 query 时已在上方走 POST :count。
    const res = await this.http.request<{ count: string | number }>(
      "GET",
      `/v1/databases/${databaseId}/collections/${collectionId}/documents:count`
    );
    return res.count ?? 0;
  }

  // 拉取集合的已提交事件流（阶段④ §4.5 补偿 API，R17 补登）：seq 升序、
  // 按当前用户可见性过滤。since_seq 为续传游标（0 = 从最老可用事件起）；
  // 游标早于重放窗口 → FailedPrecondition EVENTS.RESUME_EXPIRED（全量重拉后
  // 重新续传）。delete 事件无 data（tombstone）；按 event_id 幂等去重
  //（at-least-once）。续传**优先 next_since_seq**（R15 两级语义），仅当为 0
  // 时回退末条 change 的 seq。
  async listChanges(
    databaseId: string,
    collectionId: string,
    params?: { since_seq?: number | string; limit?: number }
  ): Promise<ListChangesResponse> {
    const query: Record<string, string | number | undefined> = {};
    if (params?.since_seq !== undefined) query.since_seq = String(params.since_seq);
    if (params?.limit !== undefined) query.limit = params.limit;
    return this.http.request<ListChangesResponse>(
      "GET",
      `/v1/databases/${databaseId}/collections/${collectionId}/changes`,
      { query }
    );
  }
}

// documentsQueryBody 组装 POST :list 的 body（Query JSON；分页字段并入——
// page_size/page_token 与 query 内同名字段以 query 为准，不重复携带）。
function documentsQueryBody(
  query: QueryAst,
  pageSize?: number,
  pageToken?: string
): Record<string, unknown> {
  const body: Record<string, unknown> = { ...query };
  if (body.pageSize === undefined && pageSize !== undefined) {
    body.pageSize = pageSize;
  }
  if (body.pageToken === undefined && pageToken !== undefined) {
    body.pageToken = pageToken;
  }
  return body;
}

File: sdk/go/client/databases.go
package client

import (
	"context"

	clientv1 "github.com/torchwooddev/torchwood/genproto/client/v1"
	sharedv1 "github.com/torchwooddev/torchwood/genproto/shared/v1"
	"google.golang.org/protobuf/types/known/structpb"
)

// DatabasesService 封装 Client API 的文档服务（动态文档 CRUD）。
// 默认数据库见 [WithDatabaseID]，可用 [Client.UseDatabase] 切换。
type DatabasesService struct {
	c  *Client
	db string
}

// newDatabasesService 构造绑定指定 db 的文档服务副本。
func newDatabasesService(c *Client, db string) *DatabasesService {
	return &DatabasesService{c: c, db: db}
}

// CreateDocument 在集合中创建文档。data 为任意 JSON 兼容结构。
func (d *DatabasesService) CreateDocument(ctx context.Context, collectionID, documentID string, data map[string]any, permissions []string) (*sharedv1.Document, error) {
	st, err := toStruct(data)
	if err != nil {
		return nil, err
	}
	return d.c.databases.CreateDocument(ctx, &clientv1.CreateDocumentRequest{
		DatabaseId:   d.db,
		CollectionId: collectionID,
		DocumentId:   documentID,
		Data:         st,
		Permissions:  permissions,
	})
}

// GetDocument 读取文档，不存在时返回 codes.NotFound。
func (d *DatabasesService) GetDocument(ctx context.Context, collectionID, documentID string) (*sharedv1.Document, error) {
	return d.c.databases.GetDocument(ctx, &clientv1.GetDocumentRequest{
		DatabaseId:   d.db,
		CollectionId: collectionID,
		DocumentId:   documentID,
	})
}

// UpdateDocument 更新文档字段；increment 对数字字段做原子增量
// （如 {"last_seq": 1}），data 与 increment 可同时使用。
// version 为用户集合 OCC 版本（GetDocument 返回的 version），必填。
func (d *DatabasesService) UpdateDocument(ctx context.Context, collectionID, documentID string, data map[string]any, increment map[string]int64, permissions []string, version int64) (*sharedv1.Document, error) {
	st, err := toStruct(data)
	if err != nil {
		return nil, err
	}
	return d.c.databases.UpdateDocument(ctx, &clientv1.UpdateDocumentRequest{
		DatabaseId:   d.db,
		CollectionId: collectionID,
		DocumentId:   documentID,
		Data:         st,
		Permissions:  permissions,
		Increment:    increment,
		Version:      &version,
	})
}

// UpsertDocument 按 conflictColumns（须匹配集合唯一索引）插入或更新文档。
func (d *DatabasesService) UpsertDocument(ctx context.Context, collectionID, documentID string, data map[string]any, conflictColumns, permissions []string) (*sharedv1.Document, error) {
	st, err := toStruct(data)
	if err != nil {
		return nil, err
	}
	return d.c.databases.UpsertDocument(ctx, &clientv1.UpsertDocumentRequest{
		DatabaseId:      d.db,
		CollectionId:    collectionID,
		DocumentId:      documentID,
		Data:            st,
		Permissions:     permissions,
		ConflictColumns: conflictColumns,
	})
}

// DeleteDocument 删除文档；version 为用户集合 OCC 版本（GetDocument 返回的 version），必填。
func (d *DatabasesService) DeleteDocument(ctx context.Context, collectionID, documentID string, version int64) error {
	_, err := d.c.databases.DeleteDocument(ctx, &clientv1.DeleteDocumentRequest{
		DatabaseId:   d.db,
		CollectionId: collectionID,
		DocumentId:   documentID,
		Version:      &version,
	})
	return err
}

// ListDocuments 按 typed AST 查询列出文档（C7 单 AST），返回文档列表与
// 下一页游标（空表示无更多）。分页走 q.PageSize/q.PageToken。
func (d *DatabasesService) ListDocuments(ctx context.Context, collectionID string, q *sharedv1.Query) ([]*sharedv1.Document, string, error) {
	req := &clientv1.ListDocumentsRequest{
		DatabaseId:   d.db,
		CollectionId: collectionID,
		Query:        q,
	}
	resp, err := d.c.databases.ListDocuments(ctx, req)
	if err != nil {
		return nil, "", err
	}
	var next string
	if resp.Meta != nil {
		next = resp.Meta.NextPageToken
	}
	return resp.Documents, next, nil
}

// CountDocuments 按 typed AST 过滤统计文档数量。
func (d *DatabasesService) CountDocuments(ctx context.Context, collectionID string, q *sharedv1.Query) (int64, error) {
	// P3-9：CountDocuments 独立 Request（不暴露无效分页参数）。
	resp, err := d.c.databases.CountDocuments(ctx, &clientv1.CountDocumentsRequest{
		DatabaseId:   d.db,
		CollectionId: collectionID,
		Query:        q,
	})
	if err != nil {
		return 0, err
	}
	return resp.Count, nil
}

// ListChanges 拉取集合的已提交事件流（阶段④ §4.5 补偿 API，语义详见
// Realtime 连接的 last_seq 续传约定）：seq 升序、按当前用户可见性过滤。
// 返回 nextSinceSeq 续传游标（R15 两级语义：满页 = 末条返回 seq；扫描
// 触顶 = 越过不可见块的扫描位置）——续传**优先使用本字段**，0 时回退
// 末条 change 的 seq；has_more=false 时恒为 0。游标早于重放窗口 →
// EVENTS.RESUME_EXPIRED（全量重拉后重新续传）。delete 事件为 tombstone。
func (d *DatabasesService) ListChanges(ctx context.Context, collectionID string, sinceSeq int64, limit int32) ([]*sharedv1.Change, bool, int64, error) {
	resp, err := d.c.databases.ListChanges(ctx, &clientv1.ListChangesRequest{
		DatabaseId:   d.db,
		CollectionId: collectionID,
		SinceSeq:     sinceSeq,
		Limit:        limit,
	})
	if err != nil {
		return nil, false, 0, err
	}
	return resp.Changes, resp.HasMore, resp.NextSinceSeq, nil
}

func toStruct(data map[string]any) (*structpb.Struct, error) {
	if len(data) == 0 {
		return nil, nil
	}
	return structpb.NewStruct(data)
}
